{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8f6806b2-a739-4ab7-8830-8b9c5767db35",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "16281389-0eb8-4dd1-8de6-1ee2ede78fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_dbNSFP(dbNSFP_data, gene, transcript):\n",
    "    map_class_dict = {\"unknown\" : 'VUS',\n",
    "                      \"Conflicting_interpretations_of_pathogenicity\":'VUS',\n",
    "                      \"not_provided\" : \"VUS\",\n",
    "                      \"not_provided\": \"VUS\",\n",
    "                      \"Uncertain_significance\" : \"VUS\",\n",
    "                      \"Pathogenic\": \"Abnormal\",\n",
    "                      \"Likely_benign\": \"Normal\",\n",
    "                      \"Benign\": \"Normal\",\n",
    "                      \"Likely_pathogenic\": \"Abnormal\",\n",
    "                      \"Pathogenic/Likely_pathogenic\": \"Abnormal\",\n",
    "                      \"Benign/Likely_benign\" : \"Normal\"}\n",
    "    dbNSFP_df = pd.read_csv(dbNSFP_data, sep='\\t')\n",
    "    required_columns = ['aaref', 'aaalt', 'aapos', 'genename', 'Ensembl_transcriptid',\n",
    "                        'clinvar_clnsig' ]\n",
    "\n",
    "    is_rankscore_column = [column for column in dbNSFP_df.columns if column.endswith('rankscore')]\n",
    "    required_columns += is_rankscore_column\n",
    "\n",
    "    dbNSFP_req_df = dbNSFP_df[required_columns]\n",
    "    dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].str.split(';')\n",
    "    dbNSFP_req_df['Ensembl_transcriptid'] = dbNSFP_req_df['Ensembl_transcriptid'].str.split(';')\n",
    "    dbNSFP_req_df['genename'] = dbNSFP_req_df['genename'].str.split(';')\n",
    "    columns_to_explode = [ 'aapos','Ensembl_transcriptid', 'genename']\n",
    "    dbNSFP_req_df = dbNSFP_req_df.apply(lambda x: x.explode() if x.name in columns_to_explode else x)\n",
    "    dbNSFP_req_df = dbNSFP_req_df[(dbNSFP_req_df['genename'] == gene) &\n",
    "                                  (dbNSFP_req_df['Ensembl_transcriptid'] == transcript)]\n",
    "\n",
    "    dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].astype(int)\n",
    "    dbNSFP_req_df = dbNSFP_req_df[dbNSFP_req_df['aapos'] >=0]\n",
    "    dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].astype(str)\n",
    "    dbNSFP_req_df['mutations'] = dbNSFP_req_df['aaref'] + dbNSFP_req_df['aapos'] + dbNSFP_req_df['aaalt']\n",
    "    \n",
    "    dbNSFP_req_df = dbNSFP_req_df.replace(\".\", np.nan)\n",
    "\n",
    "    columns_to_fillna = is_rankscore_column\n",
    "    dbNSFP_req_df[columns_to_fillna] = dbNSFP_req_df[columns_to_fillna].fillna(0)\n",
    "\n",
    "    dbNSFP_req_df['clinvar_clnsig'] = dbNSFP_req_df['clinvar_clnsig'].fillna('unknown')\n",
    "    dbNSFP_req_df['ClinVar_Class'] = dbNSFP_req_df['clinvar_clnsig'].map(map_class_dict)\n",
    "    dbNSFP_req_df = dbNSFP_req_df.drop_duplicates().reset_index(drop=True)\n",
    "    dbNSFP_req_df = dbNSFP_req_df[['mutations', 'ClinVar_Class']]\n",
    "    dbNSFP_req_df = dbNSFP_req_df[(dbNSFP_req_df['ClinVar_Class'] == \"Abnormal\") | (dbNSFP_req_df['ClinVar_Class'] == \"Normal\")]\n",
    "    dbNSFP_req_df = dbNSFP_req_df.drop_duplicates()\n",
    "    dbNSFP_req_df = dbNSFP_req_df.reset_index(drop=True)\n",
    "    return dbNSFP_req_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b37cea88-f476-40e0-912d-1d042ccafeea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_17466/900840608.py:1: DtypeWarning: Columns (0,1,7,8,9,10,11,26,49,50,52,53,54,68,101,102,104,105,150,151,152,153,154,155,156,157,158,159,162,163,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,187,188,189,192,193,194,195,196,197,198,199,200,201,202,203,205,206,207,208) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  dbNSFP_df = pd.read_csv('../dbNSFP/dbNSFP4.8a_variant.Aim2.tsv', sep='\\t')\n",
      "/tmp/ipykernel_17466/900840608.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].str.split(';')\n",
      "/tmp/ipykernel_17466/900840608.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dbNSFP_req_df['Ensembl_transcriptid'] = dbNSFP_req_df['Ensembl_transcriptid'].str.split(';')\n",
      "/tmp/ipykernel_17466/900840608.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dbNSFP_req_df['genename'] = dbNSFP_req_df['genename'].str.split(';')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>aapos</th>\n",
       "      <th>genename</th>\n",
       "      <th>Ensembl_transcriptid</th>\n",
       "      <th>clinvar_clnsig</th>\n",
       "      <th>SIFT_converted_rankscore</th>\n",
       "      <th>SIFT4G_converted_rankscore</th>\n",
       "      <th>Polyphen2_HDIV_rankscore</th>\n",
       "      <th>Polyphen2_HVAR_rankscore</th>\n",
       "      <th>...</th>\n",
       "      <th>GERP++_RS_rankscore</th>\n",
       "      <th>GERP_91_mammals_rankscore</th>\n",
       "      <th>phyloP100way_vertebrate_rankscore</th>\n",
       "      <th>phyloP470way_mammalian_rankscore</th>\n",
       "      <th>phyloP17way_primate_rankscore</th>\n",
       "      <th>phastCons100way_vertebrate_rankscore</th>\n",
       "      <th>phastCons470way_mammalian_rankscore</th>\n",
       "      <th>phastCons17way_primate_rankscore</th>\n",
       "      <th>SiPhy_29way_logOdds_rankscore</th>\n",
       "      <th>bStatistic_converted_rankscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>ABCD1</td>\n",
       "      <td>ENST00000218104</td>\n",
       "      <td>.</td>\n",
       "      <td>0.35349</td>\n",
       "      <td>0.28958</td>\n",
       "      <td>0.19712</td>\n",
       "      <td>0.12133</td>\n",
       "      <td>...</td>\n",
       "      <td>0.69188</td>\n",
       "      <td>.</td>\n",
       "      <td>0.30848</td>\n",
       "      <td>.</td>\n",
       "      <td>0.86732</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.79791</td>\n",
       "      <td>0.50529</td>\n",
       "      <td>0.97548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>ABCD1</td>\n",
       "      <td>ENST00000218104</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.49117</td>\n",
       "      <td>0.43531</td>\n",
       "      <td>0.19712</td>\n",
       "      <td>0.14941</td>\n",
       "      <td>...</td>\n",
       "      <td>0.69188</td>\n",
       "      <td>.</td>\n",
       "      <td>0.30848</td>\n",
       "      <td>.</td>\n",
       "      <td>0.86732</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.79791</td>\n",
       "      <td>0.50529</td>\n",
       "      <td>0.97548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>ABCD1</td>\n",
       "      <td>ENST00000218104</td>\n",
       "      <td>.</td>\n",
       "      <td>0.35349</td>\n",
       "      <td>0.28958</td>\n",
       "      <td>0.19712</td>\n",
       "      <td>0.12133</td>\n",
       "      <td>...</td>\n",
       "      <td>0.69188</td>\n",
       "      <td>.</td>\n",
       "      <td>0.30848</td>\n",
       "      <td>.</td>\n",
       "      <td>0.86732</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.79791</td>\n",
       "      <td>0.50529</td>\n",
       "      <td>0.97548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>K</td>\n",
       "      <td>1</td>\n",
       "      <td>ABCD1</td>\n",
       "      <td>ENST00000218104</td>\n",
       "      <td>.</td>\n",
       "      <td>0.72154</td>\n",
       "      <td>0.69154</td>\n",
       "      <td>0.39217</td>\n",
       "      <td>0.33681</td>\n",
       "      <td>...</td>\n",
       "      <td>0.69188</td>\n",
       "      <td>.</td>\n",
       "      <td>0.53</td>\n",
       "      <td>.</td>\n",
       "      <td>0.54293</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.76049</td>\n",
       "      <td>0.50529</td>\n",
       "      <td>0.97548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>ABCD1</td>\n",
       "      <td>ENST00000218104</td>\n",
       "      <td>.</td>\n",
       "      <td>0.65419</td>\n",
       "      <td>0.62352</td>\n",
       "      <td>0.35387</td>\n",
       "      <td>0.28123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.69188</td>\n",
       "      <td>.</td>\n",
       "      <td>0.53</td>\n",
       "      <td>.</td>\n",
       "      <td>0.54293</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.76049</td>\n",
       "      <td>0.50529</td>\n",
       "      <td>0.97548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450555</th>\n",
       "      <td>I</td>\n",
       "      <td>N</td>\n",
       "      <td>36</td>\n",
       "      <td>YY1</td>\n",
       "      <td>ENST00000262238</td>\n",
       "      <td>.</td>\n",
       "      <td>0.29420</td>\n",
       "      <td>0.83351</td>\n",
       "      <td>0.32022</td>\n",
       "      <td>0.32088</td>\n",
       "      <td>...</td>\n",
       "      <td>0.21832</td>\n",
       "      <td>0.18641</td>\n",
       "      <td>0.52699</td>\n",
       "      <td>0.48277</td>\n",
       "      <td>0.05231</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.32748</td>\n",
       "      <td>0.21596</td>\n",
       "      <td>0.04745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450556</th>\n",
       "      <td>I</td>\n",
       "      <td>T</td>\n",
       "      <td>36</td>\n",
       "      <td>YY1</td>\n",
       "      <td>ENST00000262238</td>\n",
       "      <td>.</td>\n",
       "      <td>0.59928</td>\n",
       "      <td>0.72224</td>\n",
       "      <td>0.07471</td>\n",
       "      <td>0.06944</td>\n",
       "      <td>...</td>\n",
       "      <td>0.21832</td>\n",
       "      <td>0.18641</td>\n",
       "      <td>0.52699</td>\n",
       "      <td>0.48277</td>\n",
       "      <td>0.05231</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.32748</td>\n",
       "      <td>0.21596</td>\n",
       "      <td>0.04745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450557</th>\n",
       "      <td>I</td>\n",
       "      <td>S</td>\n",
       "      <td>36</td>\n",
       "      <td>YY1</td>\n",
       "      <td>ENST00000262238</td>\n",
       "      <td>.</td>\n",
       "      <td>0.72154</td>\n",
       "      <td>0.79402</td>\n",
       "      <td>0.22494</td>\n",
       "      <td>0.20792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.21832</td>\n",
       "      <td>0.18641</td>\n",
       "      <td>0.52699</td>\n",
       "      <td>0.48277</td>\n",
       "      <td>0.05231</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.32748</td>\n",
       "      <td>0.21596</td>\n",
       "      <td>0.04745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450558</th>\n",
       "      <td>I</td>\n",
       "      <td>M</td>\n",
       "      <td>36</td>\n",
       "      <td>YY1</td>\n",
       "      <td>ENST00000262238</td>\n",
       "      <td>.</td>\n",
       "      <td>0.27426</td>\n",
       "      <td>0.27503</td>\n",
       "      <td>0.37037</td>\n",
       "      <td>0.39164</td>\n",
       "      <td>...</td>\n",
       "      <td>0.212</td>\n",
       "      <td>0.05632</td>\n",
       "      <td>0.34411</td>\n",
       "      <td>0.2505</td>\n",
       "      <td>0.20698</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.33024</td>\n",
       "      <td>0.04039</td>\n",
       "      <td>0.04745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450559</th>\n",
       "      <td>E</td>\n",
       "      <td>K</td>\n",
       "      <td>37</td>\n",
       "      <td>YY1</td>\n",
       "      <td>ENST00000262238</td>\n",
       "      <td>.</td>\n",
       "      <td>0.46129</td>\n",
       "      <td>0.48336</td>\n",
       "      <td>0.12183</td>\n",
       "      <td>0.08700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.18946</td>\n",
       "      <td>0.81121</td>\n",
       "      <td>0.57827</td>\n",
       "      <td>0.05248</td>\n",
       "      <td>0.71638</td>\n",
       "      <td>0.68203</td>\n",
       "      <td>0.33194</td>\n",
       "      <td>0.26294</td>\n",
       "      <td>0.04745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1975684 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       aaref aaalt aapos genename Ensembl_transcriptid clinvar_clnsig  \\\n",
       "0          M     L     1    ABCD1      ENST00000218104              .   \n",
       "1          M     V     1    ABCD1      ENST00000218104     Pathogenic   \n",
       "2          M     L     1    ABCD1      ENST00000218104              .   \n",
       "3          M     K     1    ABCD1      ENST00000218104              .   \n",
       "4          M     T     1    ABCD1      ENST00000218104              .   \n",
       "...      ...   ...   ...      ...                  ...            ...   \n",
       "450555     I     N    36      YY1      ENST00000262238              .   \n",
       "450556     I     T    36      YY1      ENST00000262238              .   \n",
       "450557     I     S    36      YY1      ENST00000262238              .   \n",
       "450558     I     M    36      YY1      ENST00000262238              .   \n",
       "450559     E     K    37      YY1      ENST00000262238              .   \n",
       "\n",
       "       SIFT_converted_rankscore SIFT4G_converted_rankscore  \\\n",
       "0                       0.35349                    0.28958   \n",
       "1                       0.49117                    0.43531   \n",
       "2                       0.35349                    0.28958   \n",
       "3                       0.72154                    0.69154   \n",
       "4                       0.65419                    0.62352   \n",
       "...                         ...                        ...   \n",
       "450555                  0.29420                    0.83351   \n",
       "450556                  0.59928                    0.72224   \n",
       "450557                  0.72154                    0.79402   \n",
       "450558                  0.27426                    0.27503   \n",
       "450559                  0.46129                    0.48336   \n",
       "\n",
       "       Polyphen2_HDIV_rankscore Polyphen2_HVAR_rankscore  ...  \\\n",
       "0                       0.19712                  0.12133  ...   \n",
       "1                       0.19712                  0.14941  ...   \n",
       "2                       0.19712                  0.12133  ...   \n",
       "3                       0.39217                  0.33681  ...   \n",
       "4                       0.35387                  0.28123  ...   \n",
       "...                         ...                      ...  ...   \n",
       "450555                  0.32022                  0.32088  ...   \n",
       "450556                  0.07471                  0.06944  ...   \n",
       "450557                  0.22494                  0.20792  ...   \n",
       "450558                  0.37037                  0.39164  ...   \n",
       "450559                  0.12183                  0.08700  ...   \n",
       "\n",
       "       GERP++_RS_rankscore GERP_91_mammals_rankscore  \\\n",
       "0                  0.69188                         .   \n",
       "1                  0.69188                         .   \n",
       "2                  0.69188                         .   \n",
       "3                  0.69188                         .   \n",
       "4                  0.69188                         .   \n",
       "...                    ...                       ...   \n",
       "450555             0.21832                   0.18641   \n",
       "450556             0.21832                   0.18641   \n",
       "450557             0.21832                   0.18641   \n",
       "450558               0.212                   0.05632   \n",
       "450559               0.203                   0.18946   \n",
       "\n",
       "       phyloP100way_vertebrate_rankscore phyloP470way_mammalian_rankscore  \\\n",
       "0                                0.30848                                .   \n",
       "1                                0.30848                                .   \n",
       "2                                0.30848                                .   \n",
       "3                                   0.53                                .   \n",
       "4                                   0.53                                .   \n",
       "...                                  ...                              ...   \n",
       "450555                           0.52699                          0.48277   \n",
       "450556                           0.52699                          0.48277   \n",
       "450557                           0.52699                          0.48277   \n",
       "450558                           0.34411                           0.2505   \n",
       "450559                           0.81121                          0.57827   \n",
       "\n",
       "       phyloP17way_primate_rankscore phastCons100way_vertebrate_rankscore  \\\n",
       "0                            0.86732                              0.71638   \n",
       "1                            0.86732                              0.71638   \n",
       "2                            0.86732                              0.71638   \n",
       "3                            0.54293                              0.71638   \n",
       "4                            0.54293                              0.71638   \n",
       "...                              ...                                  ...   \n",
       "450555                       0.05231                              0.71638   \n",
       "450556                       0.05231                              0.71638   \n",
       "450557                       0.05231                              0.71638   \n",
       "450558                       0.20698                              0.71638   \n",
       "450559                       0.05248                              0.71638   \n",
       "\n",
       "       phastCons470way_mammalian_rankscore phastCons17way_primate_rankscore  \\\n",
       "0                                  0.68203                          0.79791   \n",
       "1                                  0.68203                          0.79791   \n",
       "2                                  0.68203                          0.79791   \n",
       "3                                  0.68203                          0.76049   \n",
       "4                                  0.68203                          0.76049   \n",
       "...                                    ...                              ...   \n",
       "450555                             0.68203                          0.32748   \n",
       "450556                             0.68203                          0.32748   \n",
       "450557                             0.68203                          0.32748   \n",
       "450558                             0.68203                          0.33024   \n",
       "450559                             0.68203                          0.33194   \n",
       "\n",
       "       SiPhy_29way_logOdds_rankscore bStatistic_converted_rankscore  \n",
       "0                            0.50529                        0.97548  \n",
       "1                            0.50529                        0.97548  \n",
       "2                            0.50529                        0.97548  \n",
       "3                            0.50529                        0.97548  \n",
       "4                            0.50529                        0.97548  \n",
       "...                              ...                            ...  \n",
       "450555                       0.21596                        0.04745  \n",
       "450556                       0.21596                        0.04745  \n",
       "450557                       0.21596                        0.04745  \n",
       "450558                       0.04039                        0.04745  \n",
       "450559                       0.26294                        0.04745  \n",
       "\n",
       "[1975684 rows x 62 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbNSFP_df = pd.read_csv('../dbNSFP/dbNSFP4.8a_variant.Aim2.tsv', sep='\\t')\n",
    "required_columns = ['aaref', 'aaalt', 'aapos', 'genename', 'Ensembl_transcriptid',\n",
    "                        'clinvar_clnsig' ]\n",
    "\n",
    "is_rankscore_column = [column for column in dbNSFP_df.columns if column.endswith('rankscore')]\n",
    "required_columns += is_rankscore_column\n",
    "dbNSFP_req_df = dbNSFP_df[required_columns]\n",
    "dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].str.split(';')\n",
    "dbNSFP_req_df['Ensembl_transcriptid'] = dbNSFP_req_df['Ensembl_transcriptid'].str.split(';')\n",
    "dbNSFP_req_df['genename'] = dbNSFP_req_df['genename'].str.split(';')\n",
    "columns_to_explode = [ 'aapos','Ensembl_transcriptid', 'genename']\n",
    "dbNSFP_req_df = dbNSFP_req_df.apply(lambda x: x.explode() if x.name in columns_to_explode else x)\n",
    "#dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].astype(int)\n",
    "dbNSFP_req_df = dbNSFP_req_df[~dbNSFP_req_df['aapos'].isna()]\n",
    "dbNSFP_req_df = dbNSFP_req_df[(dbNSFP_req_df['aaref'] != 'X') | (dbNSFP_req_df['aaalt'] != 'X')]\n",
    "dbNSFP_req_df\n",
    "#dbNSFP_req_df = dbNSFP_req_df[(dbNSFP_req_df['genename'] == gene) &\n",
    "#                                  (dbNSFP_req_df['Ensembl_transcriptid'] == transcript)]\n",
    "\n",
    "#dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].astype(int)\n",
    "#dbNSFP_req_df = dbNSFP_req_df[dbNSFP_req_df['aapos'] >=0]\n",
    "#dbNSFP_req_df['aapos'] = dbNSFP_req_df['aapos'].astype(str)\n",
    "#dbNSFP_req_df['mutations'] = dbNSFP_req_df['aaref'] + dbNSFP_req_df['aapos'] + dbNSFP_req_df['aaalt']    \n",
    "#dbNSFP_req_df = dbNSFP_req_df.replace(\".\", np.nan)\n",
    "#columns_to_fillna = is_rankscore_column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9d6bceec-f72c-4e6f-a5fa-272209af4fe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list= ['CACNA1S','KDM5B','LMNA','PCSK9','CDKN1C','TNNI2','YAP1','DDX23','HNRNPA1',\n",
    "            'KRT5','TRPV4','BRCA2','PDX1','HIF1A','YY1','SMAD3','PALB2','BRCA1','CD79B',\n",
    "            'EZH1','PNPO','RAD51C','TP53','NPC1','LDLR','TNNI3','CPS1','GLS','NFE2L2',\n",
    "            'SCN2A','SCN9A','GINS1','PRNP','APP','CHEK2','CTNNB1','MCM2','STAG1','TRNT1',\n",
    "            'FGA','NSD2','PPP3CA','APC','B4GALT7','PRLR','WASF1','BRAF','EZH2','SMO','TRPV5',\n",
    "            'TRPV6','ABL1','DNM1','NOTCH1','SECISBP2','ABCD1','EMD','F9','FLNA','MECP2','NLGN3','WAS']\n",
    "dbNSFP_req_gene_df = dbNSFP_req_df[dbNSFP_req_df['genename'].isin(gene_list)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f8f717d5-56af-4ab2-9e13-487e1e0003fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "mart_df = pd.read_csv('../gene2refseq/mart_export.ensemble_refseq.mapping.tsv', sep='\\t')\n",
    "\n",
    "gene_enst_df = dbNSFP_req_gene_df[['genename', 'Ensembl_transcriptid']]\n",
    "gene_enst_df = gene_enst_df.drop_duplicates()\n",
    "gene_enst_df = gene_enst_df.reset_index(drop=True)\n",
    "gene_mart_df = pd.merge(gene_enst_df, mart_df, left_on = 'Ensembl_transcriptid', right_on = 'Transcript_stable_ID', how='left')\n",
    "gene_df = gene_mart_df[['genename']].drop_duplicates()\n",
    "gene_df =gene_df.reset_index(drop=True)\n",
    "\n",
    "gene_transcript_dict = {'NOTCH1' : 'ENST00000277541'}\n",
    "def filter_genes_with_mane(df, gene_df):\n",
    "    grouped = df.groupby('genename')\n",
    "    filtered = grouped.apply(lambda g: g if g['MANE_Select'].notna().any() else g[g['MANE_Select'].isna()].head(1))\n",
    "    filtered = filtered[['genename', 'Ensembl_transcriptid', 'MANE_Select']]\n",
    "    filtered = filtered.dropna()\n",
    "    filtered = filtered.drop_duplicates()\n",
    "    filtered = filtered.reset_index(drop=True)\n",
    "    out_df =  pd.merge(gene_df, filtered, on='genename', how='left')\n",
    "    return out_df\n",
    "\n",
    "\n",
    "filtered_df = filter_genes_with_mane(gene_mart_df, gene_df)\n",
    "\n",
    "filtered_df = filtered_df.fillna('ENST00000277541')\n",
    "enst_list = filtered_df.Ensembl_transcriptid.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "41aa16a1-54f7-492c-af79-d2d07f7cae20",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_17466/358533959.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dbNSFP_req_gene_enst_df['mutations'] = dbNSFP_req_gene_enst_df['aaref'] + dbNSFP_req_gene_enst_df['aapos'] + dbNSFP_req_gene_enst_df['aaalt']\n",
      "/tmp/ipykernel_17466/358533959.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dbNSFP_req_gene_enst_df['key'] = dbNSFP_req_gene_enst_df['mutations'] + \"_\" + dbNSFP_req_gene_enst_df['genename']\n"
     ]
    }
   ],
   "source": [
    "dbNSFP_req_gene_enst_df = dbNSFP_req_gene_df[dbNSFP_req_gene_df['Ensembl_transcriptid'].isin(enst_list)]\n",
    "dbNSFP_req_gene_enst_df['mutations'] = dbNSFP_req_gene_enst_df['aaref'] + dbNSFP_req_gene_enst_df['aapos'] + dbNSFP_req_gene_enst_df['aaalt']\n",
    "dbNSFP_req_gene_enst_df['key'] = dbNSFP_req_gene_enst_df['mutations'] + \"_\" + dbNSFP_req_gene_enst_df['genename']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b7d08972-3d0e-49f9-9fa5-9e7cb5ee655d",
   "metadata": {},
   "outputs": [],
   "source": [
    "disorder_data_df = pd.read_csv(\"../Disorder_deleterious_neutral_results.tsv\", sep='\\t')\n",
    "disorder_data_df['regions'] = 'Disorder'\n",
    "disorder_data_df['CAVA_PROTPOS']  = disorder_data_df['CAVA_PROTPOS'].astype(str)\n",
    "disorder_data_df['mutations'] = disorder_data_df['CAVA_PROTREF'] + disorder_data_df['CAVA_PROTPOS'] + disorder_data_df['CAVA_PROTALT']\n",
    "disorder_data_df['key'] = disorder_data_df['mutations'] + \"_\" +  disorder_data_df['CAVA_GENE']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d4f3ccf2-909b-48d2-89a8-3b9de6c71715",
   "metadata": {},
   "outputs": [],
   "source": [
    "disorder_data_dbnsfp_df = pd.merge(disorder_data_df, dbNSFP_req_gene_enst_df, on='key', how='left')\n",
    "keep_cols = ['mutations_x', 'GENES', 'Ensembl_transcriptid', 'Class', 'CLNSIG', \n",
    "             'SIFT_converted_rankscore',\n",
    "       'SIFT4G_converted_rankscore', 'Polyphen2_HDIV_rankscore',\n",
    "       'Polyphen2_HVAR_rankscore', 'LRT_converted_rankscore',\n",
    "       'MutationTaster_converted_rankscore', 'MutationAssessor_rankscore',\n",
    "       'FATHMM_converted_rankscore', 'PROVEAN_converted_rankscore',\n",
    "       'VEST4_rankscore', 'MetaSVM_rankscore', 'MetaLR_rankscore',\n",
    "       'MetaRNN_rankscore', 'M-CAP_rankscore', 'REVEL_rankscore',\n",
    "       'MutPred_rankscore', 'MVP_rankscore', 'gMVP_rankscore', 'MPC_rankscore',\n",
    "       'PrimateAI_rankscore', 'DEOGEN2_rankscore', 'BayesDel_addAF_rankscore',\n",
    "       'BayesDel_noAF_rankscore', 'ClinPred_rankscore', 'LIST-S2_rankscore',\n",
    "       'VARITY_R_rankscore', 'VARITY_ER_rankscore', 'VARITY_R_LOO_rankscore',\n",
    "       'VARITY_ER_LOO_rankscore', 'ESM1b_rankscore', 'EVE_rankscore',\n",
    "       'AlphaMissense_rankscore', 'PHACTboost_rankscore',\n",
    "       'MutFormer_rankscore', 'CADD_raw_rankscore', 'DANN_rankscore',\n",
    "       'fathmm-MKL_coding_rankscore', 'fathmm-XF_coding_rankscore',\n",
    "       'Eigen-raw_coding_rankscore', 'Eigen-PC-raw_coding_rankscore',\n",
    "       'GenoCanyon_rankscore', 'integrated_fitCons_rankscore',\n",
    "       'GM12878_fitCons_rankscore', 'H1-hESC_fitCons_rankscore',\n",
    "       'HUVEC_fitCons_rankscore', 'LINSIGHT_rankscore', 'GERP++_RS_rankscore',\n",
    "       'GERP_91_mammals_rankscore', 'phyloP100way_vertebrate_rankscore',\n",
    "       'phyloP470way_mammalian_rankscore', 'phyloP17way_primate_rankscore',\n",
    "       'phastCons100way_vertebrate_rankscore',\n",
    "       'phastCons470way_mammalian_rankscore',\n",
    "       'phastCons17way_primate_rankscore', 'SiPhy_29way_logOdds_rankscore',\n",
    "       'bStatistic_converted_rankscore']\n",
    "disorder_data_dbnsfp_df = disorder_data_dbnsfp_df[keep_cols]\n",
    "disorder_data_dbnsfp_df = disorder_data_dbnsfp_df.drop_duplicates()\n",
    "disorder_data_dbnsfp_df = disorder_data_dbnsfp_df.reset_index(drop=True)\n",
    "mutation_counts = disorder_data_dbnsfp_df['mutations_x'].value_counts()\n",
    "\n",
    "# Filter to get mutations with counts greater than 1\n",
    "mutations_gt_1 = mutation_counts[mutation_counts > 1].index.tolist()\n",
    "mutations_unique = mutation_counts[mutation_counts == 1].index.tolist()\n",
    "disorder_data_dbnsfp_filtered_df = disorder_data_dbnsfp_df[disorder_data_dbnsfp_df['mutations_x'].isin(mutations_unique)]\n",
    "dup_mutations_df = disorder_data_dbnsfp_df[disorder_data_dbnsfp_df['mutations_x'].isin(mutations_gt_1)]\n",
    "first_occurrences_df = dup_mutations_df.drop_duplicates(subset='mutations_x', keep='first')\n",
    "disorder_data_dbnsfp_filtered_df = pd.concat([disorder_data_dbnsfp_filtered_df,first_occurrences_df], axis=0)\n",
    "disorder_data_dbnsfp_filtered_df = disorder_data_dbnsfp_filtered_df.drop_duplicates()\n",
    "disorder_data_dbnsfp_filtered_df = disorder_data_dbnsfp_filtered_df.reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4eb5d8b4-0210-43ca-a31f-a6ee072db998",
   "metadata": {},
   "outputs": [],
   "source": [
    "disorder_data_dbnsfp_filtered_df.to_csv(\"disorder_results_with_dbNSFP.tsv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2615854a-d5c1-4097-8d6f-23de7ebe6cc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_17466/2288910870.py:7: DtypeWarning: Columns (19,20,52) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  clinvar_ordered_df = pd.read_csv('../ClinVar/clinvar.snp.header.chr.pfam.aim2genes.pathogenic_benign.cava.tsv', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CAVA_CLASS</th>\n",
       "      <th>CAVA_CSN</th>\n",
       "      <th>CAVA_GENE</th>\n",
       "      <th>CAVA_IMPACT</th>\n",
       "      <th>CAVA_LOC</th>\n",
       "      <th>...</th>\n",
       "      <th>CAVA_PROTPOS</th>\n",
       "      <th>CAVA_PROTREF</th>\n",
       "      <th>CAVA_SO</th>\n",
       "      <th>CAVA_TRANSCRIPT</th>\n",
       "      <th>CAVA_TYPE</th>\n",
       "      <th>CLNREVSTAT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>Class</th>\n",
       "      <th>mutation</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55509556</td>\n",
       "      <td>440712</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.248A&gt;C_p.Lys83Thr</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex2</td>\n",
       "      <td>...</td>\n",
       "      <td>83</td>\n",
       "      <td>K</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_174936.4</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deleterious</td>\n",
       "      <td>K83T</td>\n",
       "      <td>K83T_PCSK9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55509631</td>\n",
       "      <td>440715</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.323T&gt;C_p.Leu108Pro</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex2</td>\n",
       "      <td>...</td>\n",
       "      <td>108</td>\n",
       "      <td>L</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_174936.4</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deleterious</td>\n",
       "      <td>L108P</td>\n",
       "      <td>L108P_PCSK9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55518073</td>\n",
       "      <td>2874</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.646T&gt;C_p.Phe216Leu</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex4</td>\n",
       "      <td>...</td>\n",
       "      <td>216</td>\n",
       "      <td>F</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_174936.4</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deleterious</td>\n",
       "      <td>F216L</td>\n",
       "      <td>F216L_PCSK9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55523127</td>\n",
       "      <td>265939</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.1120G&gt;C_p.Asp374His</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex7</td>\n",
       "      <td>...</td>\n",
       "      <td>374</td>\n",
       "      <td>D</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_174936.4</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>criteria_provided,_multiple_submitters,_no_con...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deleterious</td>\n",
       "      <td>D374H</td>\n",
       "      <td>D374H_PCSK9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55523127</td>\n",
       "      <td>2875</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.1120G&gt;T_p.Asp374Tyr</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex7</td>\n",
       "      <td>...</td>\n",
       "      <td>374</td>\n",
       "      <td>D</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_174936.4</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>criteria_provided,_multiple_submitters,_no_con...</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>Deleterious</td>\n",
       "      <td>D374Y</td>\n",
       "      <td>D374Y_PCSK9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1735</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153594724</td>\n",
       "      <td>1429176</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.1180A&gt;G_p.Ser394Gly</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex8</td>\n",
       "      <td>...</td>\n",
       "      <td>394</td>\n",
       "      <td>S</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_001110556.2</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Neutral</td>\n",
       "      <td>S394G</td>\n",
       "      <td>S394G_FLNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1736</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153594965</td>\n",
       "      <td>282366</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.1030G&gt;A_p.Val344Ile</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex7</td>\n",
       "      <td>...</td>\n",
       "      <td>344</td>\n",
       "      <td>V</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_001110556.2</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>criteria_provided,_multiple_submitters,_no_con...</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>Neutral</td>\n",
       "      <td>V344I</td>\n",
       "      <td>V344I_FLNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153594986</td>\n",
       "      <td>1003574</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.1009G&gt;A_p.Asp337Asn</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex7</td>\n",
       "      <td>...</td>\n",
       "      <td>337</td>\n",
       "      <td>D</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_001110556.2</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Neutral</td>\n",
       "      <td>D337N</td>\n",
       "      <td>D337N_FLNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1738</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153595129</td>\n",
       "      <td>701764</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.958G&gt;A_p.Val320Met</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex6</td>\n",
       "      <td>...</td>\n",
       "      <td>320</td>\n",
       "      <td>V</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_001110556.2</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>criteria_provided,_multiple_submitters,_no_con...</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>Neutral</td>\n",
       "      <td>V320M</td>\n",
       "      <td>V320M_FLNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153596077</td>\n",
       "      <td>1396503</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NSY</td>\n",
       "      <td>c.652G&gt;A_p.Ala218Thr</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>2</td>\n",
       "      <td>Ex4</td>\n",
       "      <td>...</td>\n",
       "      <td>218</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NM_001110556.2</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Neutral</td>\n",
       "      <td>A218T</td>\n",
       "      <td>A218T_FLNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1740 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM        POS       ID REF ALT CAVA_CLASS               CAVA_CSN  \\\n",
       "0      chr1   55509556   440712   A   C        NSY    c.248A>C_p.Lys83Thr   \n",
       "1      chr1   55509631   440715   T   C        NSY   c.323T>C_p.Leu108Pro   \n",
       "2      chr1   55518073     2874   T   C        NSY   c.646T>C_p.Phe216Leu   \n",
       "3      chr1   55523127   265939   G   C        NSY  c.1120G>C_p.Asp374His   \n",
       "4      chr1   55523127     2875   G   T        NSY  c.1120G>T_p.Asp374Tyr   \n",
       "...     ...        ...      ...  ..  ..        ...                    ...   \n",
       "1735   chrX  153594724  1429176   T   C        NSY  c.1180A>G_p.Ser394Gly   \n",
       "1736   chrX  153594965   282366   C   T        NSY  c.1030G>A_p.Val344Ile   \n",
       "1737   chrX  153594986  1003574   C   T        NSY  c.1009G>A_p.Asp337Asn   \n",
       "1738   chrX  153595129   701764   C   T        NSY   c.958G>A_p.Val320Met   \n",
       "1739   chrX  153596077  1396503   C   T        NSY   c.652G>A_p.Ala218Thr   \n",
       "\n",
       "     CAVA_GENE CAVA_IMPACT CAVA_LOC  ... CAVA_PROTPOS CAVA_PROTREF  \\\n",
       "0        PCSK9           2      Ex2  ...           83            K   \n",
       "1        PCSK9           2      Ex2  ...          108            L   \n",
       "2        PCSK9           2      Ex4  ...          216            F   \n",
       "3        PCSK9           2      Ex7  ...          374            D   \n",
       "4        PCSK9           2      Ex7  ...          374            D   \n",
       "...        ...         ...      ...  ...          ...          ...   \n",
       "1735      FLNA           2      Ex8  ...          394            S   \n",
       "1736      FLNA           2      Ex7  ...          344            V   \n",
       "1737      FLNA           2      Ex7  ...          337            D   \n",
       "1738      FLNA           2      Ex6  ...          320            V   \n",
       "1739      FLNA           2      Ex4  ...          218            A   \n",
       "\n",
       "               CAVA_SO CAVA_TRANSCRIPT     CAVA_TYPE  \\\n",
       "0     missense_variant     NM_174936.4  Substitution   \n",
       "1     missense_variant     NM_174936.4  Substitution   \n",
       "2     missense_variant     NM_174936.4  Substitution   \n",
       "3     missense_variant     NM_174936.4  Substitution   \n",
       "4     missense_variant     NM_174936.4  Substitution   \n",
       "...                ...             ...           ...   \n",
       "1735  missense_variant  NM_001110556.2  Substitution   \n",
       "1736  missense_variant  NM_001110556.2  Substitution   \n",
       "1737  missense_variant  NM_001110556.2  Substitution   \n",
       "1738  missense_variant  NM_001110556.2  Substitution   \n",
       "1739  missense_variant  NM_001110556.2  Substitution   \n",
       "\n",
       "                                             CLNREVSTAT  \\\n",
       "0                        no_assertion_criteria_provided   \n",
       "1                        no_assertion_criteria_provided   \n",
       "2                        no_assertion_criteria_provided   \n",
       "3     criteria_provided,_multiple_submitters,_no_con...   \n",
       "4     criteria_provided,_multiple_submitters,_no_con...   \n",
       "...                                                 ...   \n",
       "1735                criteria_provided,_single_submitter   \n",
       "1736  criteria_provided,_multiple_submitters,_no_con...   \n",
       "1737                criteria_provided,_single_submitter   \n",
       "1738  criteria_provided,_multiple_submitters,_no_con...   \n",
       "1739                criteria_provided,_single_submitter   \n",
       "\n",
       "                            CLNSIG        Class mutation          key  \n",
       "0                       Pathogenic  Deleterious     K83T   K83T_PCSK9  \n",
       "1                       Pathogenic  Deleterious    L108P  L108P_PCSK9  \n",
       "2                       Pathogenic  Deleterious    F216L  F216L_PCSK9  \n",
       "3                       Pathogenic  Deleterious    D374H  D374H_PCSK9  \n",
       "4     Pathogenic/Likely_pathogenic  Deleterious    D374Y  D374Y_PCSK9  \n",
       "...                            ...          ...      ...          ...  \n",
       "1735                        Benign      Neutral    S394G   S394G_FLNA  \n",
       "1736          Benign/Likely_benign      Neutral    V344I   V344I_FLNA  \n",
       "1737                        Benign      Neutral    D337N   D337N_FLNA  \n",
       "1738          Benign/Likely_benign      Neutral    V320M   V320M_FLNA  \n",
       "1739                        Benign      Neutral    A218T   A218T_FLNA  \n",
       "\n",
       "[1740 rows x 21 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_class_label(x):\n",
    "    del_class = ['Pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic/Likely_pathogenic/Pathogenic,_low_penetrance', 'Pathogenic|risk_factor']\n",
    "    if x in del_class:\n",
    "        return 'Deleterious'\n",
    "    else:\n",
    "        return \"Neutral\"\n",
    "clinvar_ordered_df = pd.read_csv('../ClinVar/clinvar.snp.header.chr.pfam.aim2genes.pathogenic_benign.cava.tsv', sep='\\t')\n",
    "#print(clinvar_ordered_df.columns)\n",
    "drop_colmns = ['DBVARID', 'GENEINFO', 'MC',\n",
    "       'ONC', 'ONCCONF', 'ONCDISDB', 'ONCDISDBINCL', 'ONCDN', 'ONCDNINCL',\n",
    "       'ONCINCL', 'ONCREVSTAT', 'ORIGIN', 'RS', 'SCI', 'SCIDISDB',\n",
    "       'SCIDISDBINCL', 'SCIDN', 'SCIDNINCL', 'SCIINCL', 'SCIREVSTAT', \n",
    "               'CLNDN', 'CLNDNINCL', 'CLNDNINCL', 'CLNHGVS', 'CLNVCSO', 'CLNVI',\n",
    "              'QUAL', 'FILTER', 'AF_ESP', 'SCIDISDBINCL', 'SCIDN', 'SCIDNINCL', 'SCIINCL', 'SCIREVSTAT',\n",
    "       'AF_EXAC', 'AF_TGP', 'ALLELEID', 'CLNSIGCONF',\n",
    "       'CLNSIGINCL', 'CLNVC', 'CLNVCSO', 'CLNVI', 'DBVARID', 'GENEINFO', 'MC',\n",
    "       'ONC', 'ONCCONF', 'ONCDISDB', 'ONCDISDBINCL', 'ONCDN', 'ONCDNINCL',\n",
    "       'ONCINCL', 'ONCREVSTAT', 'ORIGIN', 'RS', 'SCI', 'SCIDISDB',\n",
    "       'SCIDISDBINCL', 'SCIDN', 'SCIDNINCL', 'SCIINCL', 'SCIREVSTAT', 'CAVA_ALTANN', 'CAVA_GENEID',\n",
    "               'CAVA_ALTCLASS', 'CAVA_ALTFLAG', 'CLNDISDBINCL', 'CAVA_DBSNP', 'CAVA_ALTSO', 'CLNDISDB', 'CAVA_TRINFO']\n",
    "clinvar_ordered_df = clinvar_ordered_df.drop(drop_colmns, axis=1)\n",
    "clinvar_ordered_df = clinvar_ordered_df[clinvar_ordered_df['CAVA_SO'] == 'missense_variant']\n",
    "clinvar_ordered_df = clinvar_ordered_df.drop_duplicates()\n",
    "clinvar_ordered_df = clinvar_ordered_df.reset_index(drop=True)\n",
    "clinvar_ordered_df['Class' ] = clinvar_ordered_df['CLNSIG'].apply(lambda x: create_class_label(x))\n",
    "clinvar_ordered_df['mutation'] = clinvar_ordered_df['CAVA_PROTREF'] + clinvar_ordered_df['CAVA_PROTPOS'] + clinvar_ordered_df['CAVA_PROTALT']\n",
    "clinvar_ordered_df['key'] = clinvar_ordered_df['mutation'] + '_' + clinvar_ordered_df['CAVA_GENE']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a363481a-600f-4fa0-94c0-87294aa7a507",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_ordered_dbNSFP_df = pd.merge(clinvar_ordered_df, dbNSFP_req_gene_enst_df, on='key', how='left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4744fb79-5916-4146-9789-11bd7a8eeeee",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_ordered_dbNSFP_df = clinvar_ordered_dbNSFP_df[~clinvar_ordered_dbNSFP_df['phyloP470way_mammalian_rankscore'].isna()]\n",
    "clinvar_ordered_dbNSFP_df.to_csv(\"ordered_regions_results_with_dbNSFP.tsv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b6e4b52-a716-44a4-8fb8-e770d4e647c4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
